File: phi_plugin_koishi/lib/GameRecord.ts
import LevelRecord from './LevelRecord'
import ByteReader from './ByteReader';
import Util from './Util'
import getInfo from '../model/getInfo'
import { idString } from '../model/type/type';

export default interface GameRecord {
    name: string;
    version: number;
    data: ByteReader;
    Record: any;
    songsnum: number;
    init(err: Array<string>): any;
}

export default class GameRecord {
    static name = "gameRecord";
    static version = 1;
    constructor(data: string) {
        this.name = "gameRecord";
        this.version = 1;
        this.data = new ByteReader(data)
        this.Record = {}
        this.songsnum = 0
    }

    /**
     * 
     * @param {Array} err 错误消息
     */
    async init(err: string[]) {
        this.songsnum = this.data.getVarInt()
        while (this.data.remaining() > 0) {
            let key = this.data.getString().replace(/\.0$/, '') as idString;
            this.data.skipVarInt()
            let length = this.data.getByte();
            let fc = this.data.getByte();
            let song = [];


            for (let level = 0; level < 5; level++) {
                if (Util.getBit(length, level)) {
                    song[level] = new LevelRecord();
                    song[level].score = this.data.getInt();
                    song[level].acc = this.data.getFloat();
                    song[level].fc = (song[level].score == 1000000 && song[level].acc == 100) ? true : Util.getBit(fc, level);

                }
            }
            if (!getInfo.idgetsong(key)) {
                err.push(key)
            }
            this.Record[key] = song
        }
    }
}

File: phi_plugin_js/lib/GameRecord.py
from typing import Any

from zhenxun.services.log import logger

from .ByteReader import ByteReader
from .LevelRecord import LevelRecord


class GameRecord:
    """游戏记录"""

    name: str = "gameRecord"
    version: int = 1

    def __init__(self, data: bytes) -> None:
        """
        初始化

        :param data: 二进制数据
        """
        self.Record: dict[str, list[LevelRecord | None]] = {}
        self.data = data

    async def init(self, song_data: list[Any]) -> None:
        """
        初始化记录

        :param song_data: 歌曲数据
        """
        try:
            reader = ByteReader(self.data)
            if reader.getByte() != self.version:
                logger.error("版本号已更新，请更新PhigrosLibrary。")
                raise ValueError("版本号已更新")

            while reader.remaining() > 0:
                key = reader.getString()
                song: list[LevelRecord | None] = []
                for _ in range(4):
                    if reader.getVarInt() == 1:
                        record = LevelRecord()
                        record.fc = bool(reader.getVarInt())
                        record.score = reader.getVarInt()
                        record.acc = reader.getFloat()
                        song.append(record)
                    else:
                        song.append(None)
                self.Record[key] = song

        except Exception as e:
            logger.error("初始化记录失败", e=e)
            raise ValueError("初始化记录失败") from e
